### PR TITLE
feat: Minkowski penalties

### PR DESCRIPTION
**Pull request by jiriminarcik**
# Description

This PR reimplements the `disjoint` function in `packages/core/src/contrib/Constraints.ts` for rectangles and convex polygons. The disjoint constraint for shapes `A` and `B` is implemented as a distance from the origin to the Minkowski sum of `A` and `-B`. The mathematical approach is described in more detail [here](https://github.com/penrose/penrose/files/7810485/minkowski-rectangles-refs.pdf).

Example of optimization for axis-aligned rectangles:

Example of optimization for polygons (custom convex partitioning was used for the static shape):

This implementation also works for 1-simplexes (line segments):

# Implementation strategy and design decisions

The current implementation supports **convex polygons only**. The support for non-convex polygons can be added by implementing a convex partitioning algorithm in the function `convexPartitions` in `packages/core/src/contrib/Constraints.ts`.

This approach can lead to large computational graphs, which cause slow code generation step of the autodiff engine (see #652 for more details). The full computational graphs and their properties for two basic interactions are listed below.

Two axis-aligned rectangles (187 nodes, 269 edges and tree depth is 21):
rectangles-graph

Two triangles (866 nodes, 1312 edges and tree depth is 35):
triangles-graph

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `yarn test`
- [x] I ran `yarn docs` and there were no errors when generating the HTML site
- [x] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

### CODE REVIEW

**Comment by wodeni** [comment]
@jiriminarcik do you mind pasting in paths to examples from this branch? Would be very helpful if you can point us to where you were having performance problems with the autodiff codegen.

**Review by wodeni — COMMENTED** [review]
Overall looks great to me! I haven't looked at the theory. It would be good if someone (maybe @keenancrane?) can have a look at the math here. Just two comments on AD/perf below.

File: packages/core/src/contrib/Constraints.ts [hunk 1]
```diff
@@ -485,13 +499,21 @@ export const constrDict = {
...
-      return mul(overlapX, overlapY);
+      const [pc1, pc2] = rectangleDifference(box1, box2);
+      const [xp, yp] = ops.vmul(varOf(0.5), ops.vadd(pc1, pc2));
+      const [xr, yr] = ops.vmul(varOf(0.5), ops.vsub(pc2, pc1));
+      const [xq, yq] = ops.vsub([absVal(xp), absVal(yp)], [xr, yr]);
+      const e1 = sqrt(
+        add(
+          varOf(10e-15),
+          add(
+            squared(max(sub(xp, xr), varOf(0.0))),
```
[the remark this answers is not in the document]
Comment (jiriminarcik) [reply packages/core/src/contrib/Constraints.ts:510]: As far as I can see, there is no difference in the functionality.. And it seems to be used interchangeably across the codebase. But I'll change it to `constOf` everywhere I have an explicit number inside.

File: packages/core/src/contrib/Constraints.ts [hunk 2]
```diff
@@ -1021,3 +1043,121 @@ export const areDisjointBoxes = (a: BBox.BBox, b: BBox.BBox): VarAD => {
...
+  const [xb1, xb2, yb1, yb2] = [
+    BBox.minX(box2),
+    BBox.maxX(box2),
+    BBox.minY(box2),
+    BBox.maxY(box2),
+  ];
+  // Compute coordinates of the new rectangle
+  const xs = [sub(xa1, xb1), sub(xa2, xb2), sub(xa1, xb2), sub(xa2, xb1)];
+  const ys = [sub(ya1, yb1), sub(ya2, yb2), sub(ya1, yb2), sub(ya2, yb1)];
+  const xc1 = xs.reduce((x, y) => min(x, y, true));
```
[the remark this answers is not in the document]
Comment (jiriminarcik) [reply packages/core/src/contrib/Constraints.ts]: Added `minN` and `maxN`. This also led to some readability improvements.

**Comment by keenancrane** [comment]
Looks great to me!

The PDF has a reference to [4], which seems to be missing.  (But probably fine to go ahead with the PR…)

**Comment by jiriminarcik** [comment]
> Looks great to me!
> 
> The PDF has a reference to [4], which seems to be missing. (But probably fine to go ahead with the PR…)

fixed

**Pull request merged**